lib/plugins/package/lib/zipService.js

```diff
@@ -14,39 +14,9 @@ const _ = require('lodash');
 const excludeNodeDevDependenciesMemoized = _.memoize(excludeNodeDevDependencies);
 
 module.exports = {
-  async zipService(exclude, include, zipFileName) {
-    const params = await this.excludeDevDependencies({
-      exclude,
-      include,
-      zipFileName,
-    });
-    return this.zip(params);
-  },
-
-  async excludeDevDependencies(params) {
-    const servicePath = this.serverless.config.servicePath;
-
-    let excludeDevDependencies = this.serverless.service.package.excludeDevDependencies;
-    if (excludeDevDependencies === undefined || excludeDevDependencies === null) {
-      excludeDevDependencies = true;
-    }
-
-    if (excludeDevDependencies) {
-      this.serverless.cli.log('Excluding development dependencies...');
-
-      const exAndInNode = await excludeNodeDevDependenciesMemoized(servicePath);
-      params.exclude = _.union(params.exclude, exAndInNode.exclude); //eslint-disable-line
-      params.include = _.union(params.include, exAndInNode.include); //eslint-disable-line
-      return params;
-    }
-
-    return params;
-  },
-
-  zip(params) {
-    return this.resolveFilePathsFromPatterns(params).then(filePaths =>
-      this.zipFiles(filePaths, params.zipFileName)
-    );
+  async excludeDevDependencies() {
```

Comment:
This `excludeDevDependencies ` method received the `package` config and updated the `exclude` property with the dev dependencies.

* It shouldn't be in `zipService` but in `packageService` (I didn't change it to don't make the review even more complicated.)
* The method is really tightly coupled with the include/exclude format. Now, it simple returns the list of devDependencies' patterns so the calling method can add them to the patterns in any way (future-proffness).

Reply:
Thanks for clarifications. Actually `excludeDevDependencies` doesn't really receive `package` config directly.

The actual packaging config is consumed in two places:
- https://github.com/serverless/serverless/blob/854526816265a425d2d3fe7ae4285b242893b434/lib/plugins/package/lib/packageService.js#L21
- https://github.com/serverless/serverless/blob/854526816265a425d2d3fe7ae4285b242893b434/lib/plugins/package/lib/packageService.js#L197-L201

Then `this.getIncludes` and `this.getExcludes` resolve internal _include_ and _exclude_ patterns collections (which are not `package` config objects) which are processed further.

Technically to implement this functionality, all we need is to tweak `this.getIncludes` so it combines `package.patterns` into returned _include_ patterns, and otherwise nothing else is needed to be done I think.

Of course it'll be also nice to clear internals so they also process single _patterns_ collections and not both _include_ and _exclude_, but as it can be seen a separate concern, and it's best if it's addressed with other PR (ideally PR's should be minimal and focus on one thing, especially that we deal here with really sensitive functionality).

Additionally we have a refactor planned to how we resolve files in a service and dev dependencies: https://github.com/serverless/serverless/issues/8586 and that will replace some of the logic in `zipService` and `packageService`, so it may not be worth to imply any refactors to it at this point